### DISCUSSION TITLE
Can you help me to figure out how to connect two many-to-many models (Posts and Tags) in Prisma and Nexus?

### DISCUSSION THREAD

**Discussion by lumenwrites** [prisma/prisma#7985] [Q&A] [answered]
Hi! I want to create a post with a list of tags attached to it. The models are connected many-to-many (one post can have several tags, and one tag can have several posts in it). 

Here are my prisma models:
```prisma
    model Post {
      id String @id @default(cuid())
      slug String @unique
      title String
      body String
      tags Tag[]
    }
    
    model Tag {
      id String @id @default(cuid())
      posts Post[]
      name String
      slug String @unique
    }
```
And here's a mutation where I'm trying to create a post, and attach tags to it:
```ts
    t.field('createPost', {
      type: 'Post',
      args: {
        title: nonNull(stringArg()),
        body: stringArg(),
        tags: list(arg({ type: 'TagInput' }))
      },
      resolve: async (_, args, context: Context) => {
        // Create tags if they don't exist
        const tags = await Promise.all(
          args.tags.map((tag) =>
            context.prisma.tag.upsert({
              create: omit(tag, "id"),
              update: tag,
              where: { id: tag.id || "" },
            })
          )
        )
        return context.prisma.post.create({
          data: {
            title: args.title,
            body: args.body,
            slug: `${slugify(args.title)}-${cuid()}`,
            tags: {
              //disconnect: [How do I disconnect already existing tags???]
              connect: [{id:"ckql6n0i40000of9yzi6d8bv5"}]
            },
            authorId: getUserId(context),
            published: true, // make it false once Edit post works.
          },
        })
      },
    })
```
This doesn't seem to be working. 

I'm getting an error:

    Invalid `prisma.post.create()` invocation:
    {
      data: {
        title: 'Post with tags',
        body: 'Post with tags body',
        slug: 'Post-with-tags-ckql7jy850003uz9y8xri51zf',
        tags: {
          connect: [
            {
              id: 'ckql6n0i40000of9yzi6d8bv5'
            }
          ]
        },
      }
    }
    Unknown arg `tags` in data.tags for type PostUncheckedCreateInput. Available args:
    type PostUncheckedCreateInput {
      id?: String
      title: String
      body: String
      slug: String
    }

It seems like the `tags` field on the post is missing? But I did run `prisma generate` and `prisma migrate`, what else could be causing this issue?

My second question is - how do I replace already connected tags with the new ones? For example, in the `updatePost` mutation, some of the tags may have existed before the update, but then the user wants to remove them from the post. How do I correctly replace all the tags with the new list? If I should first `disconnect` all the old tags and then `connect` the new ones - is there a way to do it all in the same mutation?

**Comment by ryands17** [comment, marked as answer]
@lumenwrites 👋 

Your code should look like this:

```ts
return context.prisma.post.create({
          data: {
            title: args.title,
            body: args.body,
            slug: `${slugify(args.title)}-${cuid()}`,
            tags: {
              connect: [{id:"ckql6n0i40000of9yzi6d8bv5"}]
            },
            author: { connect: { id: getUserId(context) } },
            published: true, // make it false once Edit post works.
          },
})
```

And then it will work fine. Also why are you using `disconnect` on newly created posts? That doesn't make sense.

**Reply by lumenwrites** [reply]
Unfortunately that doesn't solve the issue, I'm still getting the same error. I do have the author and the published field, and I'm setting them just as you described, I've just removed them from my question to make it more concise and avoid confusing people with extra unrelated fields.

I don't think this has anything to do with the author field. If I comment out the tags, everything works. For some reason Prisma doesn't seem to know that the Post has `tags` field on it when I'm doing the mutation. But queries work and creating tags works fine.

**Reply by ryands17** [reply]
There was a typo. Updated my answer, it is related to that as you need to use `connect` and cannot directly set the `authorId`.

**Reply by lumenwrites** [reply]
Nope, unfortunately the same exact error if I use connect like this:

            tags: {
              connect: [{id:"ckqldp05l0003ae9y8v8i6tdk"}]
            },

Or create like this:

            tags: {
              create: {
                    name: "test",
                    slug: "test"
              }
            },

Can you please take a look at my code and see if I'm doing anything wrong?

- Prisma schema is [here](https://github.com/lumenwrites/sketchdaily/blob/master/backend/prisma/schema.prisma).
- `createPost` mutation is [here](https://github.com/lumenwrites/sketchdaily/blob/master/backend/nexus/postMutations.ts).

Just to clarify - setting the author works just fine, I'm pretty sure this issue has nothing to do with the author field at all.

**Reply by ryands17** [reply]
Here's the working copy of the code. You can run `yarn seed` or `npm run seed` and see that all the relations work fine.
[test-prisma.zip](https://github.com/prisma/prisma/files/6752539/test-prisma.zip)

**Reply by lumenwrites** [reply]
Thank you very much! Turns out using the tags like this:

```
        tags: {
          connect: [{id:"ckqldp05l0003ae9y8v8i6tdk"}]
        },
```

was the correct way after all.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
